### PR TITLE
Add type reduction support to Min, Max and Pow

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add type reduction support to Min, Max and Pow
Update the C++ type reduction infrastructure to allow specifying an opset for the supported types list, as those can change across opset versions.
Minor updates to the type usage tracking script

**Motivation and Context**
Reduce build size for 1P models

### CODE REVIEW

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 1]
```diff
@@ -197,11 +204,12 @@ struct EnabledTypes {
...
+ * @param OpSet The opset to use for the supported types list.
  * @param ArgDirection Direction of the given Op kernel argument - Input or Output.
  * @param ArgIndex Index of the given Op kernel argument.
  */
 #define ORT_OP_KERNEL_ARG_ENABLED_TYPE_TUPLE(                                       \
-    OpProvider, OpDomain, OpName, ArgDirection, ArgIndex)                           \
+    OpProvider, OpDomain, OpName, Opset, ArgDirection, ArgIndex)                    \
   ::boost::mp11::mp_rename<                                                         \
       ORT_OP_KERNEL_ARG_ENABLED_TYPE_LIST(                                          \
           OpProvider, OpDomain, OpName, ArgDirection, ArgIndex, SupportedTypeList), \
```
Comment (edgchen1) [inline onnxruntime/core/providers/op_kernel_type_control.h:215]: OpSet should be added to the arguments here too.
or we could remove this macro if there's not a need for it. we can change the other one to return std::tuple if we want to use std types.

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc [hunk 2]
```diff
@@ -30,8 +30,9 @@ using namespace boost::mp11;
 namespace onnxruntime {
 
 namespace op_kernel_type_control {
+// we're using one set of types for all opsets of Cast, so use -1 as a generic opset value
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/cast_op.cc:33]: would it be useful to have -1 as a constant for a generic opset? i suppose any value would work, but that may help with consistency in the future.

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/op_kernel_type_control.h:215]: Will add for now. If we convert all the ops for the next release and don't use this macro we could remove it at that point.

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 3]
```diff
@@ -151,62 +155,119 @@ struct EnabledTypes {
...
  * @param ArgDirection Direction of the given Op kernel argument - Input or Output.
  * @param ArgIndex Index of the given Op kernel argument.
  */
-#define ORT_OP_KERNEL_ARG_ENABLED_TYPE_TUPLE(                                       \
-    OpProvider, OpDomain, OpName, ArgDirection, ArgIndex)                           \
-  ::boost::mp11::mp_rename<                                                         \
-      ORT_OP_KERNEL_ARG_ENABLED_TYPE_LIST(                                          \
-          OpProvider, OpDomain, OpName, ArgDirection, ArgIndex, SupportedTypeList), \
+#define ORT_OP_KERNEL_ARG_ENABLED_TYPE_TUPLE(                                              \
+    OpProvider, OpDomain, OpName, Opset, ArgDirection, ArgIndex)                           \
```
Comment (edgchen1) [inline onnxruntime/core/providers/op_kernel_type_control.h:251]: "Opset" -> "OpSet"

**Review by edgchen1 — APPROVED** [review]
LGTM

**Comment by skottmckay** [comment]
>     #   - Priority 1P models

Test

---
Refers to: tools/python/util/ort_format_model/operator_type_usage_processors.py:240 in 7e2dd16. [](commit_id = 7e2dd1602dce8a64fb441800f645d9ef49fec9f0, deletion_comment = False)

**Pull request merged**